### ISSUE TITLE
Big difference between training with 1 GPU vs 8 GPUs (via DDP) 

### ISSUE THREAD

**Issue by PabloPeso** [speechbrain/speechbrain#899]
Hi all,

I am using the recipe LibriSpeech/ASR/transformer with the hyperparameters saved in hparams/conformer_small.yaml.

I firstly tried training on a single GPU with the following result
```
epoch: 1, lr: 1.76e-04, steps: 4394, optimizer: Adam - train loss: 2.61e+02 - valid loss: 1.41e+02, valid ACC: 1.86e-01
epoch: 2, lr: 3.51e-04, steps: 8788, optimizer: Adam - train loss: 1.48e+02 - valid loss: 1.11e+02, valid ACC: 3.76e-01
epoch: 3, lr: 5.27e-04, steps: 13182, optimizer: Adam - train loss: 67.33 - valid loss: 58.62, valid ACC: 6.64e-01
epoch: 4, lr: 7.03e-04, steps: 17576, optimizer: Adam - train loss: 49.27 - valid loss: 75.51, valid ACC: 6.06e-01
epoch: 5, lr: 8.79e-04, steps: 21970, optimizer: Adam - train loss: 41.83 - valid loss: 21.69, valid ACC: 8.57e-01
epoch: 6, lr: 9.74e-04, steps: 26364, optimizer: Adam - train loss: 37.18 - valid loss: 14.90, valid ACC: 8.99e-01
epoch: 7, lr: 9.02e-04, steps: 30758, optimizer: Adam - train loss: 32.34 - valid loss: 12.56, valid ACC: 9.14e-01
epoch: 8, lr: 8.43e-04, steps: 35152, optimizer: Adam - train loss: 28.62 - valid loss: 15.06, valid ACC: 8.97e-01
epoch: 9, lr: 7.95e-04, steps: 39546, optimizer: Adam - train loss: 25.86 - valid loss: 10.85, valid ACC: 9.24e-01
epoch: 10, lr: 7.54e-04, steps: 43940, optimizer: Adam - train loss: 23.81 - valid loss: 10.45, valid ACC: 9.28e-01, valid WER: 7.79
```

which makes sense to me. However, when I move to 8 GPUs (and changing the `gradient_accumulation` from 4 to 1 and reducing `batch_size` from 16 to 8 in order to get the same global batch size as the case of 1 GPU) by running `python -m torch.distributed.launch --nproc_per_node=8 train.py hparams/conformer_small.yaml --distributed_launch --distributed_backend=nccl` I get

```
epoch: 1, lr: 1.76e-04, steps: 4395, optimizer: Adam - train loss: 2.76e+02 - valid loss: 1.55e+02, valid ACC: 1.50e-01
epoch: 2, lr: 3.52e-04, steps: 8790, optimizer: Adam - train loss: 2.41e+02 - valid loss: 1.49e+02, valid ACC: 1.72e-01
epoch: 3, lr: 5.27e-04, steps: 13185, optimizer: Adam - train loss: 2.35e+02 - valid loss: 1.48e+02, valid ACC: 1.83e-01
epoch: 4, lr: 7.03e-04, steps: 17580, optimizer: Adam - train loss: 2.30e+02 - valid loss: 1.49e+02, valid ACC: 1.87e-01
epoch: 5, lr: 8.79e-04, steps: 21975, optimizer: Adam - train loss: 2.27e+02 - valid loss: 1.49e+02, valid ACC: 1.94e-01
epoch: 6, lr: 9.74e-04, steps: 26370, optimizer: Adam - train loss: 2.25e+02 - valid loss: 1.49e+02, valid ACC: 1.96e-01
epoch: 7, lr: 9.01e-04, steps: 30765, optimizer: Adam - train loss: 2.22e+02 - valid loss: 1.49e+02, valid ACC: 2.01e-01
epoch: 8, lr: 8.43e-04, steps: 35160, optimizer: Adam - train loss: 2.18e+02 - valid loss: 1.50e+02, valid ACC: 2.03e-01
epoch: 9, lr: 7.95e-04, steps: 39555, optimizer: Adam - train loss: 2.15e+02 - valid loss: 1.52e+02, valid ACC: 2.03e-01
epoch: 10, lr: 7.54e-04, steps: 43950, optimizer: Adam - train loss: 2.13e+02 - valid loss: 1.54e+02, valid ACC: 2.02e-01, valid WER: 2.58e+02
epoch: 11, lr: 7.19e-04, steps: 48345, optimizer: Adam - train loss: 2.11e+02 - valid loss: 1.54e+02, valid ACC: 2.00e-01
epoch: 12, lr: 6.89e-04, steps: 52740, optimizer: Adam - train loss: 2.09e+02 - valid loss: 1.55e+02, valid ACC: 1.98e-01
epoch: 13, lr: 6.61e-04, steps: 57135, optimizer: Adam - train loss: 2.07e+02 - valid loss: 1.56e+02, valid ACC: 1.97e-01
epoch: 14, lr: 6.37e-04, steps: 61530, optimizer: Adam - train loss: 2.06e+02 - valid loss: 1.56e+02, valid ACC: 1.95e-01
epoch: 15, lr: 6.16e-04, steps: 65925, optimizer: Adam - train loss: 2.04e+02 - valid loss: 1.57e+02, valid ACC: 1.93e-01
epoch: 16, lr: 5.96e-04, steps: 70320, optimizer: Adam - train loss: 2.03e+02 - valid loss: 1.58e+02, valid ACC: 1.91e-01
epoch: 17, lr: 5.78e-04, steps: 74715, optimizer: Adam - train loss: 2.02e+02 - valid loss: 1.58e+02, valid ACC: 1.89e-01
epoch: 18, lr: 5.62e-04, steps: 79110, optimizer: Adam - train loss: 2.01e+02 - valid loss: 1.59e+02, valid ACC: 1.87e-01
epoch: 19, lr: 5.47e-04, steps: 83505, optimizer: Adam - train loss: 2.00e+02 - valid loss: 1.57e+02, valid ACC: 1.87e-01
epoch: 20, lr: 5.33e-04, steps: 87900, optimizer: Adam - train loss: 1.99e+02 - valid loss: 1.59e+02, valid ACC: 1.85e-01, valid WER: 1.97e+02
epoch: 21, lr: 5.20e-04, steps: 92295, optimizer: Adam - train loss: 1.99e+02 - valid loss: 1.58e+02, valid ACC: 1.84e-01
epoch: 22, lr: 5.08e-04, steps: 96690, optimizer: Adam - train loss: 1.98e+02 - valid loss: 1.59e+02, valid ACC: 1.83e-01
```
which seems to indicate that there is something wrong because the train_loss seems to get stuck. I tried also to increase the global batch size from 64 to 128 by increasing the batch size to 16 but I get similar outcome. Is this a problem related to DDP and therefore an expected outcome?

Thanks for your help.

**Comment by TParcollet** [comment]
Hi, we experienced this issue in the past, but we weren't unable to find the reason ... We could use some help if you find the time to debug. We will try to find someone to work on that again very soon.

**Comment by TParcollet** [comment]
One possible issue is the synchronization that may not occur at the accum level. I don't remember exactly how DDP manages to synchronize the gradients, I have to check again. But maybe your setup isn't identical, and you would need to do the same accum that with one GPU but reducing the batch size.

**Comment by PabloPeso** [comment]
Hi,

Thanks for your answer.

I have just tried with the same gradient accumulation, i.e. 4, and therefore reducing the batch size to 2 to keep the global size to the same level, i.e. 4 x 2 x 8 = 64. However I get very similar behaviour
```
epoch: 1, lr: 1.76e-04, steps: 4394, optimizer: Adam - train loss: 2.85e+02 - valid loss: 1.55e+02, valid ACC: 1.51e-01
epoch: 2, lr: 3.51e-04, steps: 8788, optimizer: Adam - train loss: 2.41e+02 - valid loss: 1.49e+02, valid ACC: 1.73e-01
epoch: 3, lr: 5.27e-04, steps: 13182, optimizer: Adam - train loss: 2.34e+02 - valid loss: 1.47e+02, valid ACC: 1.83e-01
epoch: 4, lr: 7.03e-04, steps: 17576, optimizer: Adam - train loss: 2.29e+02 - valid loss: 1.45e+02, valid ACC: 1.89e-01
epoch: 5, lr: 8.79e-04, steps: 21970, optimizer: Adam - train loss: 2.26e+02 - valid loss: 1.45e+02, valid ACC: 1.94e-01
epoch: 6, lr: 9.74e-04, steps: 26364, optimizer: Adam - train loss: 2.24e+02 - valid loss: 1.43e+02, valid ACC: 1.97e-01
epoch: 7, lr: 9.02e-04, steps: 30758, optimizer: Adam - train loss: 2.20e+02 - valid loss: 1.44e+02, valid ACC: 2.01e-01
epoch: 8, lr: 8.43e-04, steps: 35152, optimizer: Adam - train loss: 2.18e+02 - valid loss: 1.45e+02, valid ACC: 2.02e-01
epoch: 9, lr: 7.95e-04, steps: 39546, optimizer: Adam - train loss: 2.14e+02 - valid loss: 1.44e+02, valid ACC: 2.03e-01
epoch: 10, lr: 7.54e-04, steps: 43940, optimizer: Adam - train loss: 2.13e+02 - valid loss: 1.48e+02, valid ACC: 2.02e-01, valid WER: 2.13e+02
epoch: 11, lr: 7.19e-04, steps: 48334, optimizer: Adam - train loss: 2.10e+02 - valid loss: 1.51e+02, valid ACC: 2.02e-01
epoch: 12, lr: 6.89e-04, steps: 52728, optimizer: Adam - train loss: 2.09e+02 - valid loss: 1.53e+02, valid ACC: 2.00e-01
epoch: 13, lr: 6.62e-04, steps: 57122, optimizer: Adam - train loss: 2.07e+02 - valid loss: 1.53e+02, valid ACC: 1.98e-01
epoch: 14, lr: 6.37e-04, steps: 61516, optimizer: Adam - train loss: 2.06e+02 - valid loss: 1.49e+02, valid ACC: 1.96e-01
epoch: 15, lr: 6.16e-04, steps: 65910, optimizer: Adam - train loss: 2.04e+02 - valid loss: 1.55e+02, valid ACC: 1.94e-01
epoch: 16, lr: 5.96e-04, steps: 70304, optimizer: Adam - train loss: 2.03e+02 - valid loss: 1.53e+02, valid ACC: 1.93e-01
epoch: 17, lr: 5.79e-04, steps: 74698, optimizer: Adam - train loss: 2.02e+02 - valid loss: 1.56e+02, valid ACC: 1.91e-01
```

What would you think it would be the most obvious cause for this behaviour? Is this related with the issue with DP not able to run on multiple GPU https://github.com/speechbrain/speechbrain/issues/868? Could it be entirely related to the transformer implementation itself? Let me know your thoughts to debug this.

Cheers.

**Comment by TParcollet** [comment]
I think there is something weird going on with the transformer. Because I believe that these behaviors are only being observed with them. Transformers are being refactored, but it could make sense to identify the issue here. Transformers are also the only recipes that use gradient accumulation, so it could be related to that.

**Comment by PabloPeso** [comment]
Hi @TParcollet,

Could it be due to the fact that transformers heavily rely on LayerNorm (as opposed to BatchNorm) and we are not synchronizing the learnable parameters of these layers (gamma and beta) from each GPU?

For BatchNorm I can see in the code ` module = SyncBatchNorm.convert_sync_batchnorm(module)`, but I think this only works for BatchNorm layers, right?

Do you think this could be the issue with transformers on multi-GPU?

Cheers,

**Comment by TParcollet** [comment]
I honestly don't really know for now, but yes, this could be a first approach.

**Comment by danpovey** [comment]
i dont think layernorm needs synchronization.  the bias and scale are just
normal parameters.
I think this is a general transformer thing.
plus librispeech utts are.long, so hard to learn alignment at.first.

**Comment by TParcollet** [comment]
@PabloPeso indeed, layer norm statistics are layer-dependent, so they do not really need to be synchronized in between GPUs. I am trying to take a look into this issue.

**Comment by PabloPeso** [comment]
Thanks @danpovey.
I see the issue you mention, however if it was a general transformer thing... shouldn't we see the convergence issue also in the case of running on a single GPU? Because at the moment on a single GPU I can see the training loss decreases as expected, however for 8 GPUs the training loss seems to get stuck.

**Comment by PabloPeso** [comment]
Thanks @TParcollet for helping. Let me know if I can help in any way.

**Comment by TParcollet** [comment]
We spotted few errors in the recipe that could cause issues with multi-GPU. Let us work a bit on that part :)

**Comment by mravanelli** [comment]
I think this is fixed, right @TParcollet? But apparently we still have to merge the PR with that fix. I suggest doing it after the final checks.

**Comment by TParcollet** [comment]
Yes, PR is pending. But I think there is still a problem with DP ... unsure though, this is something @popcornell mentionned.

**Comment by PabloPeso** [comment]
I have tested this PR and here is the result
```
epoch: 1, lr: 2.19e-05, steps: 549, optimizer: Adam - train loss: 5.07e+02 - valid loss: 2.35e+02, valid ACC: 5.71e-02
epoch: 2, lr: 4.39e-05, steps: 1098, optimizer: Adam - train loss: 2.89e+02 - valid loss: 1.62e+02, valid ACC: 1.36e-01
epoch: 3, lr: 6.58e-05, steps: 1647, optimizer: Adam - train loss: 2.52e+02 - valid loss: 1.54e+02, valid ACC: 1.56e-01
epoch: 4, lr: 8.78e-05, steps: 2196, optimizer: Adam - train loss: 2.42e+02 - valid loss: 1.50e+02, valid ACC: 1.69e-01
epoch: 5, lr: 1.10e-04, steps: 2745, optimizer: Adam - train loss: 2.34e+02 - valid loss: 1.45e+02, valid ACC: 1.82e-01
epoch: 6, lr: 1.32e-04, steps: 3294, optimizer: Adam - train loss: 2.25e+02 - valid loss: 1.37e+02, valid ACC: 2.02e-01
epoch: 7, lr: 1.54e-04, steps: 3843, optimizer: Adam - train loss: 2.11e+02 - valid loss: 1.27e+02, valid ACC: 2.18e-01
epoch: 8, lr: 1.76e-04, steps: 4392, optimizer: Adam - train loss: 1.92e+02 - valid loss: 1.20e+02, valid ACC: 2.28e-01
epoch: 9, lr: 1.98e-04, steps: 4941, optimizer: Adam - train loss: 1.56e+02 - valid loss: 1.17e+02, valid ACC: 2.55e-01
epoch: 10, lr: 2.20e-04, steps: 5490, optimizer: Adam - train loss: 1.23e+02 - valid loss: 1.04e+02, valid ACC: 3.45e-01, valid WER: 72.81
```
It seems that it improves over previous 8GPU training, but still far from 1GPU.

By the way, I had to create by hand the links to the LM and tokenizer because I was having the following issue:
`requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://huggingface.co//gpfswork/rech/kxg/uyk21ll/datasets/huggingface_hub/asr-transformer-transformerlm-librispeech/resolve/main/lm.ckpt`

**Comment by TParcollet** [comment]
I don't understand, I tried with 6 GPUs, it worked well ... Let me try again.

**Comment by TParcollet** [comment]
Ok @PabloPeso I think you have some hyper parameters issues when scaling to 8 GPUs. See the number of steps in the log you posted ? The last one is at 5490 (backward steps, basically). Your results at this number of steps are comparable with the ones obtained with a single GPU, no ?

**Comment by mravanelli** [comment]
@PabloPeso, any news? I think @TParcollet is not observing anymore problems with that, right?

**Comment by dophist** [comment]
I think the problem has been fixed, sharing some numbers relating to this:

repo head commit: 6cbef2 (Aug 17, 2021)

AISHELL-1 transformer recipe:

* 1 GPU (batch_size = 8, grad_accu = 4)
```
epoch: 1, lr: 5.93e-05, steps: 3753, optimizer: Adam - train loss: 1.03e+02 - valid loss: 81.78, valid ACC: 2.33e-01
epoch: 2, lr: 1.19e-04, steps: 7506, optimizer: Adam - train loss: 77.92 - valid loss: 75.48, valid ACC: 2.69e-01
epoch: 3, lr: 1.78e-04, steps: 11259, optimizer: Adam - train loss: 66.95 - valid loss: 49.64, valid ACC: 4.19e-01
```
* 4 GPUs with DDP invocation (batch_size = 8, grad_accu = 1)
```
epoch: 1, lr: 5.93e-05, steps: 3754, optimizer: Adam - train loss: 1.03e+02 - valid loss: 81.94, valid ACC: 2.31e-01
epoch: 2, lr: 1.19e-04, steps: 7508, optimizer: Adam - train loss: 78.03 - valid loss: 75.58, valid ACC: 2.66e-01
epoch: 3, lr: 1.78e-04, steps: 11262, optimizer: Adam - train loss: 67.15 - valid loss: 50.36, valid ACC: 4.12e-01
```
Reasonable differences.

**Comment by PabloPeso** [comment]
I won't be able to test this in the coming two weeks. I can let you know
the results in 3 weeks time if that's still helpful. Thanks.

**Comment by TParcollet** [comment]
Ok, let's close the issue for now. @PabloPeso feel free to reopen it depending on your results.

**Comment by PabloPeso** [comment]
Writing this comment just to help closing this issue.

I have now tested it with the right backward steps per epoch and it seems pretty close to the 1 GPU case. In order to have the same steps when using 8 GPUs I have set `batch_size: 8` and `gradient_accumulation: 1`. Here are the results:
```
epoch: 1, lr: 1.76e-04, steps: 4395, optimizer: Adam - train loss: 2.61e+02 - valid loss: 1.43e+02, valid ACC: 1.85e-01
epoch: 2, lr: 3.52e-04, steps: 8790, optimizer: Adam - train loss: 1.55e+02 - valid loss: 94.45, valid ACC: 4.31e-01
epoch: 3, lr: 5.27e-04, steps: 13185, optimizer: Adam - train loss: 71.98 - valid loss: 44.09, valid ACC: 7.05e-01
epoch: 4, lr: 7.03e-04, steps: 17580, optimizer: Adam - train loss: 52.36 - valid loss: 38.74, valid ACC: 7.63e-01
epoch: 5, lr: 8.79e-04, steps: 21975, optimizer: Adam - train loss: 44.02 - valid loss: 22.74, valid ACC: 8.50e-01
epoch: 6, lr: 9.74e-04, steps: 26370, optimizer: Adam - train loss: 38.70 - valid loss: 16.57, valid ACC: 8.87e-01
epoch: 7, lr: 9.01e-04, steps: 30765, optimizer: Adam - train loss: 33.46 - valid loss: 14.13, valid ACC: 9.06e-01
epoch: 8, lr: 8.43e-04, steps: 35160, optimizer: Adam - train loss: 29.12 - valid loss: 15.11, valid ACC: 8.98e-01
epoch: 9, lr: 7.95e-04, steps: 39555, optimizer: Adam - train loss: 26.34 - valid loss: 13.50, valid ACC: 9.10e-01
epoch: 10, lr: 7.54e-04, steps: 43950, optimizer: Adam - train loss: 24.53 - valid loss: 12.55, valid ACC: 9.15e-01, valid WER: 8.61
```

This is very close to initial results I included in the first comment for 1 GPU (copying results below)
```
epoch: 1, lr: 1.76e-04, steps: 4394, optimizer: Adam - train loss: 2.61e+02 - valid loss: 1.41e+02, valid ACC: 1.86e-01
epoch: 2, lr: 3.51e-04, steps: 8788, optimizer: Adam - train loss: 1.48e+02 - valid loss: 1.11e+02, valid ACC: 3.76e-01
epoch: 3, lr: 5.27e-04, steps: 13182, optimizer: Adam - train loss: 67.33 - valid loss: 58.62, valid ACC: 6.64e-01
epoch: 4, lr: 7.03e-04, steps: 17576, optimizer: Adam - train loss: 49.27 - valid loss: 75.51, valid ACC: 6.06e-01
epoch: 5, lr: 8.79e-04, steps: 21970, optimizer: Adam - train loss: 41.83 - valid loss: 21.69, valid ACC: 8.57e-01
epoch: 6, lr: 9.74e-04, steps: 26364, optimizer: Adam - train loss: 37.18 - valid loss: 14.90, valid ACC: 8.99e-01
epoch: 7, lr: 9.02e-04, steps: 30758, optimizer: Adam - train loss: 32.34 - valid loss: 12.56, valid ACC: 9.14e-01
epoch: 8, lr: 8.43e-04, steps: 35152, optimizer: Adam - train loss: 28.62 - valid loss: 15.06, valid ACC: 8.97e-01
epoch: 9, lr: 7.95e-04, steps: 39546, optimizer: Adam - train loss: 25.86 - valid loss: 10.85, valid ACC: 9.24e-01
epoch: 10, lr: 7.54e-04, steps: 43940, optimizer: Adam - train loss: 23.81 - valid loss: 10.45, valid ACC: 9.28e-01, valid WER: 7.79
```

Thanks for your help.
